README.md
# Generate SVG Aruco Marker Images with JavaScript

`aruco-marker` is a JavaScript library that can generate marker images (fiducials) for the [Aruco](https://github.com/paroj/aruco) augmented reality marker library. Aruco codes can be recognized by the original Aruco C++ library, or in the browser by [js-aruco](https://github.com/jcmellado/js-aruco). `aruco-marker` generates images as SVG, making them easy to scale to any size or print out. There is also a custom HTML element allowing you to easily embed codes anywhere on a page.

[![NPM version](https://badge.fury.io/js/aruco-marker.svg)](https://www.npmjs.com/package/aruco-marker)

# Demos

There are two demos that show off marker generation either directly or via the custom HTML element:

- [Random Aruco markers via custom element](http://bhollis.github.io/aruco-marker/demos/element.html) ([View Source](https://github.com/bhollis/aruco-marker/blob/master/demos/element.html))
- [Random Aruco markers via string](http://bhollis.github.io/aruco-marker/demos/index.html) ([View Source](https://github.com/bhollis/aruco-marker/blob/master/demos/index.html))

# Usage

`aruco-marker` is available for use in the browser, or in NodeJS. It is installable as `aruco-marker` from NPM.

```javascript
import { arucoToSVGString } from 'aruco-marker';

const svgImage = arucoToSVGString(155, '500px'); // the size is optional
document.getElementById('marker').innerHTML = svgImage;
```

See [`demos/index.html`](https://github.com/bhollis/aruco-marker/blob/master/demos/index.html) for a complete example.

# Custom Element

`aruco-marker` is available for use in the browser, or in NodeJS. It is installable as `aruco-marker` from NPM.

```html
<script src="https://unpkg.com/aruco-marker/element"></script>

<aruco-marker markerid="155" size="500px"></aruco-marker>
```

# SVG to Canvas

While SVG images are very flexible and are perfect for most applications, you can use [drawImage](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage) to draw an SVG image to an HTML Canvas in order to obtain an image data URL if you need a raster image.

# Developing

First, install NodeJS however you like for your system (on macOS, I use `brew install node`).

Then check out and build the project:

```bash
git clone https://github.com/bhollis/aruco-marker
cd aruco-marker
npm ci
npm run demo
```

## License

Copyright (c) 2014 Benjamin Hollis. MIT Licensed, see [MIT-LICENSE.txt](https://github.com/bhollis/aruco-marker/blob/master/MIT-LICENSE.txt) for details.


CHANGELOG.md
3.0.0
-----

* The library now exports an ESM-compatible arucoToSVGString function rather than an object-based API.
* Added an HTML custom element `<aruco-marker markerid="155"></aruco-marker>` that works regardless of framework.
* Removed AngularJS component.
* Removed AMD and Bower support.

2.0.0
-----

* Fix marker orientation - it was diagonally inverted!
* Add CommonJS and AMD support to the Angular directive.

1.0.0
-----

* Initial release


test/arucoMarkerSpec.js
describe("ArucoMarker", function() {
  var ArucoMarker = require('../aruco-marker');

  it("accepts a valid ID", function() {
    new ArucoMarker(15);
  });

  it("rejects a negative ID", function() {
    expect(function() {
      new ArucoMarker(-15);
    }).toThrow();
  });

  // it("rejects a too-large ID", function() {
  //   expect(function() {
  //     new ArucoMarker(10000);
  //   }).toThrow();
  // });

  it("can generate a marker matrix for 1", function() {
    var expected = [[1, 1, 1, 1, 1],
                    [0, 0, 0, 0, 0],
                    [0, 0, 0, 0, 1],
                    [0, 0, 0, 0, 1],
                    [0, 0, 0, 0, 1]];
    var marker = new ArucoMarker(1);

    expect(marker.markerMatrix()).toEqual(expected);
  });

  it("can generate a marker matrix for 18", function() {
    var expected = [[1, 1, 1, 1, 0],
                    [0, 0, 0, 0, 1],
                    [0, 0, 1, 0, 0],
                    [0, 0, 1, 0, 0],
                    [0, 0, 1, 0, 1]];
    var marker = new ArucoMarker(18);

    expect(marker.markerMatrix()).toEqual(expected);
  });

  it("can generate SVG", function() {
    var marker = new ArucoMarker(18);

    expect(marker.toSVG()).toEqual(
      '<svg  viewBox="0 0 7 7" version="1.1" xmlns="http://www.w3.org/2000/svg">\n' +
      '  <rect x="0" y="0" width="7" height="7" fill="black"/>\n' +
      '  <rect x="1" y="1" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="1" y="2" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="1" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="3" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="4" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="5" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="1" y="4" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="2" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="5" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '</svg>');
  });

  it("can generate SVG with dimensions", function() {
    var marker = new ArucoMarker(18);

    expect(marker.toSVG('5em')).toEqual(
      '<svg height="5em" width="5em" viewBox="0 0 7 7" version="1.1" xmlns="http://www.w3.org/2000/svg">\n' +
      '  <rect x="0" y="0" width="7" height="7" fill="black"/>\n' +
      '  <rect x="1" y="1" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="1" y="2" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="1" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="3" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="4" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="5" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="1" y="4" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="2" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '  <rect x="5" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
      '</svg>');
  });

});


src/aruco-marker.test.ts
import { arucoMarkerMatrix, arucoToSVGString } from './aruco-marker.js';
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

describe('arucoMarkerMatrix', () => {
	it('accepts a valid ID', () => {
		arucoMarkerMatrix(15);
	});

	it('rejects a negative ID', () => {
		assert.throws(() => {
			arucoMarkerMatrix(-15);
		});
	});

	it('rejects a too-large ID', () => {
		assert.throws(() => {
			arucoMarkerMatrix(10000);
		});
	});

	it('can generate a marker matrix for 1', () => {
		var expected = [
			[1, 1, 1, 1, 1],
			[0, 0, 0, 0, 0],
			[0, 0, 0, 0, 1],
			[0, 0, 0, 0, 1],
			[0, 0, 0, 0, 1],
		];

		assert.deepEqual(expected, arucoMarkerMatrix(1));
	});

	it('can generate a marker matrix for 18', () => {
		var expected = [
			[1, 1, 1, 1, 0],
			[0, 0, 0, 0, 1],
			[0, 0, 1, 0, 0],
			[0, 0, 1, 0, 0],
			[0, 0, 1, 0, 1],
		];

		assert.deepEqual(expected, arucoMarkerMatrix(18));
	});

	it('can generate SVG', () => {
		const expected =
			'<svg  viewBox="0 0 7 7" version="1.1" xmlns="http://www.w3.org/2000/svg">\n' +
			'  <rect x="0" y="0" width="7" height="7" fill="black"/>\n' +
			'  <rect x="1" y="1" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="1" y="2" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="1" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="3" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="4" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="5" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="1" y="4" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="2" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="5" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'</svg>';

		assert.equal(expected, arucoToSVGString(18));
	});

	it('can generate SVG with dimensions', () => {
		const expected =
			'<svg height="5em" width="5em" viewBox="0 0 7 7" version="1.1" xmlns="http://www.w3.org/2000/svg">\n' +
			'  <rect x="0" y="0" width="7" height="7" fill="black"/>\n' +
			'  <rect x="1" y="1" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="1" y="2" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="1" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="3" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="4" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="5" y="3" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="1" y="4" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="2" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'  <rect x="5" y="5" width="1" height="1" fill="white" stroke="white" stroke-width="0.01" />\n' +
			'</svg>';

		assert.equal(expected, arucoToSVGString(18, '5em'));
	});
});


Gruntfile.js
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    banner: '/*! <%= pkg.name %> <%= pkg.version %> <%= grunt.template.today("yyyy-mm-dd") %> - MIT Licensed, see http://github.com/bhollis/aruco-marker */\n',
    copy: {
      main: {
        src: ['<%= pkg.name %>.js', 'ng-<%= pkg.name %>.js'],
        dest: 'build/'
      },
    },
    uglify: {
      build: {
        files: {
          'build/<%= pkg.name %>.min.js': '<%= pkg.name %>.js',
          'build/ng-<%= pkg.name %>.min.js': 'ng-<%= pkg.name %>.js'
        }
      }
    },
    usebanner: {
      dist: {
        options: {
          position: 'top',
          banner: '<%= banner %>',
          linebreak: false,
        },
        files: {
          src: [ 'build/*.js' ]
        }
      }
    },
    jshint: {
      // define the files to lint
      files: ['Gruntfile.js', '<%= pkg.name %>.js', 'ng-<%= pkg.name %>.js', 'test/**/*.js'],
      // configure JSHint (documented at http://www.jshint.com/docs/)
      options: {
        // more options here if you want to override JSHint defaults
        globals: {
          module: true
        }
      }
    },
    jasmine_node: {
      projectRoot: ".",
      verbose: true,
      forceExit: true,
      jUnit: {
        report: false,
        savePath : "./build/reports/jasmine/",
        useDotNotation: true,
        consolidate: true
      }
    },
    watch: {
      files: ['<%= jshint.files %>'],
      tasks: ['test']
    }
  });

  grunt.loadNpmTasks('grunt-contrib-copy');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-jasmine-node');
  grunt.loadNpmTasks('grunt-banner');

  grunt.registerTask('test', ['jshint', 'jasmine_node']);
  grunt.registerTask('default', ['jshint', 'test', 'copy', 'uglify', 'usebanner']);
};


src/element.ts
import { arucoToSVGString } from 'aruco-marker';

class ArucoMarkerElement extends HTMLElement {
	constructor() {
		super();
	}

	static get observedAttributes() {
		return ['markerid', 'size'];
	}

	connectedCallback() {
		if (!this.shadowRoot) {
			this.attachShadow({ mode: 'open' });
			this._upgradeProperty('markerId');
			this._upgradeProperty('size');
		}

		const svgContents = arucoToSVGString(this.markerId, this.size ?? undefined);
		this.shadowRoot!.innerHTML = svgContents;
	}

	_upgradeProperty(prop: 'markerId' | 'size') {
		if (this.hasOwnProperty(prop)) {
			let value = this[prop];
			delete this[prop];
			// @ts-expect-error ts(2322)
			this[prop] = value;
		}
	}

	attributeChangedCallback() {
		this.connectedCallback();
	}

	public get markerId() {
		const markerId = parseInt(this.getAttribute('markerid') ?? '', 10);
		if (Number.isNaN(markerId)) {
			throw new Error('markerid attribute must be set');
		}
		return markerId;
	}

	public set markerId(value: number) {
		this.setAttribute('markerid', value.toString());
	}

	public get size() {
		return this.getAttribute('size');
	}

	public set size(value: string | null) {
		if (!value) {
			this.removeAttribute('size');
		} else {
			this.setAttribute('size', value);
		}
	}
}

customElements.define('aruco-marker', ArucoMarkerElement, {});


ng-aruco-marker.js
// AngularJS Directive for Aruco marker images
// Usage:
// <aruco-marker marker-id="{{markerId}}" size="500px"/>
(function (root, factory) {
  if (typeof module !== 'undefined' && module.exports) {
    // CommonJS
    if (typeof angular === 'undefined') {
      factory(require('angular'));
    } else {
      factory(angular);
    }
    module.exports = 'ngDialog';
  } else if (typeof define === 'function' && define.amd) {
    // AMD
    define(['angular'], factory);
  } else {
    // Global Variables
    factory(root.angular);
  }
}(this, function (angular) {
  angular.module('arucoMarker', []).directive('arucoMarker', function() {
    return {
      restrict: 'E',
      scope: {
        markerId: '@',
        size: '@'
      },
      link: function(scope, element, attrs) {
        scope.$watch('markerId', function(newVal, oldVal) {
          var marker = new ArucoMarker(+scope.markerId);
          element.html(marker.toSVG(scope.size));
        });
      }
    };
  });
}));


src/aruco-marker.ts
// Based on https://github.com/rmsalinas/aruco/blob/master/trunk/src/arucofidmarkers.cpp

/** Generate a marker as a 5x5 matrix of 0s and 1s. */
export function arucoMarkerMatrix(id: number) {
	if (id < 0 || id > 1023) {
		throw new RangeError('Marker ID must be in the range [0..1023]');
	}

	const ids = [16, 23, 9, 14];
	let index = 0,
		val = 0,
		x = 0,
		y = 0;
	const marker = [
		[0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0],
	];

	for (y = 0; y < 5; y++) {
		index = (id >> (2 * (4 - y))) & 3;
		val = ids[index];
		for (x = 0; x < 5; x++) {
			marker[x][y] = (val >> (4 - x)) & 1;
		}
	}

	return marker;
}

/**
 * Create an SVG image of the marker, as a string.
 * Optionally pass a size (in any SVG-compatible units) or leave it out to size it on your own.
 */
export function arucoToSVGString(id: number, size?: number | string) {
	let x = 0,
		y = 0;
	const marker = arucoMarkerMatrix(id);
	let image = '';

	if (size) {
		size = 'height="' + size + '" width="' + size + '"';
	} else {
		size = '';
	}

	image =
		'<svg ' +
		size +
		' viewBox="0 0 7 7" version="1.1" xmlns="http://www.w3.org/2000/svg">\n' +
		'  <rect x="0" y="0" width="7" height="7" fill="black"/>\n';

	for (y = 0; y < 5; y++) {
		for (x = 0; x < 5; x++) {
			if (marker[x][y] === 1) {
				image +=
					'  <rect x="' +
					(x + 1) +
					'" y="' +
					(y + 1) +
					'" width="1" height="1" fill="white" ' +
					// Slight stroke to get around aliasing issues with adjacent rectangles
					'stroke="white" stroke-width="0.01" />\n';
			}
		}
	}

	image += '</svg>';

	return image;
}


demos/demo-server.js
import * as fs from 'node:fs';
import * as http from 'node:http';
import * as path from 'node:path';

const PORT = 8000;

const MIME_TYPES = {
	default: 'application/octet-stream',
	html: 'text/html; charset=UTF-8',
	js: 'application/javascript',
	css: 'text/css',
	png: 'image/png',
	jpg: 'image/jpg',
	gif: 'image/gif',
	ico: 'image/x-icon',
	svg: 'image/svg+xml',
};

const STATIC_PATHS = [process.cwd()];

const toBool = [() => true, () => false];

const prepareFile = async (url) => {
	for (const STATIC_PATH of STATIC_PATHS) {
		const paths = [STATIC_PATH, url];
		if (url.endsWith('/')) paths.push('index.html');
		const filePath = path.join(...paths);
		const pathTraversal = !filePath.startsWith(STATIC_PATH);
		const exists = await fs.promises.access(filePath).then(...toBool);
		const found = !pathTraversal && exists;
		if (!found) {
			continue;
		}
		const ext = path.extname(filePath).substring(1).toLowerCase();
		const stream = fs.createReadStream(filePath);
		return { found, ext, stream };
	}
	return {
		found: false,
		ext: 'html',
		stream: fs.createReadStream(path.join(process.cwd(), './demos/404.html')),
	};
};

http
	.createServer(async (req, res) => {
		const file = await prepareFile(req.url);
		const statusCode = file.found ? 200 : 404;
		const mimeType = MIME_TYPES[file.ext] || MIME_TYPES.default;
		res.writeHead(statusCode, { 'Content-Type': mimeType });
		file.stream.pipe(res);
		console.log(`${req.method} ${req.url} ${statusCode}`);
	})
	.listen(PORT);

console.log(`Server running at http://127.0.0.1:${PORT}/demos/`);


aruco-marker.js
// Export for use via AMD, Node.js, or a browser global.
// See https://github.com/umdjs/umd/blob/master/returnExportsGlobal.js
(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define([], function () {
      return (root.ArucoMarker = factory());
    });
  } else if (typeof exports === 'object') {
    // Node. Does not work with strict CommonJS, but
    // only CommonJS-like enviroments that support module.exports,
    // like Node.
    module.exports = factory();
  } else {
    // Browser globals
    root.ArucoMarker = factory();
  }
}(this, function () {

  // Create an ArucoMarker object by its ID, which can then be used to generate images.
  // The id must be in the range [0..1023]
  // Based on https://github.com/rmsalinas/aruco/blob/master/trunk/src/arucofidmarkers.cpp
  function ArucoMarker(id) {
    if (id < 0 || id > 33554432) {
      throw new RangeError('Marker ID must be a postitive number');
    }

    this.id = id;
  }

  ArucoMarker.prototype = {
    // Generate a marker as a 5x5 matrix of 0s and 1s.
    markerMatrix: function() {
      var ids = [16, 23, 9, 14];
      var index, val, x, y;
      var marker = [[0, 0, 0, 0, 0],
                 [0, 0, 0, 0, 0],
                 [0, 0, 0, 0, 0],
                 [0, 0, 0, 0, 0],
                 [0, 0, 0, 0, 0]];

      for (y = 0; y < 5; y++) {
        index = (this.id >> 2 * (4 - y)) & 3;
        val = ids[index];
        for (x = 0; x < 5; x++) {
          if ((val >> (4 - x)) & 1) {
            marker[x][y] = 1;
          } else {
            marker[x][y] = 0;
          }
        }
      }

      return marker;
    },

    // Create an SVG image of the marker, as a string.
    // Optionally pass a size (in any SVG-compatible units) or leave it out to size it on your own.
    toSVG: function(size) {
      var x, y;
      var marker = this.markerMatrix();
      var image;

      if (size) {
        size = 'height="' + size + '" width="' + size + '"';
      } else {
        size = '';
      }

      image = '<svg ' + size + ' viewBox="0 0 7 7" version="1.1" xmlns="http://www.w3.org/2000/svg">\n' +
        '  <rect x="0" y="0" width="7" height="7" fill="black"/>\n';

      for (y = 0; y < 5; y++) {
        for (x = 0; x < 5; x++) {
          if (marker[x][y] === 1) {
            image += '  <rect x="' + (x + 1) + '" y="' + (y + 1) +
              '" width="1" height="1" fill="white" ' +
              // Slight stroke to get around aliasing issues with adjacent rectangles
              'stroke="white" stroke-width="0.01" />\n';
          }
        }
      }

      image += '</svg>';

      return image;
    }
  };

  return ArucoMarker;
}));
